File: itc/files/C12/ensemble.py
from random import randint
from time import time
import sys

sys.setrecursionlimit(10**6)

def appartient(x,ens):
    for i in ens:
        if i == x:
            return True
    return False

def intersection(ens1, ens2):
    res = []
    for i in ens1:
        if appartient(i, ens2):
            res.append(i)
    return res

def intersection2(ens1, ens2):
    if ens1==[] or ens2==[]:
        return []
    else:
        if ens1[0]==ens2[0]:
            return [ens1[0]] + intersection2(ens1[1:], ens2[1:])
        elif ens1[0]<ens2[0]:
            return intersection2(ens1[1:], ens2)
        else:
            return intersection2(ens1, ens2[1:])

n = 5000
ens1 = sorted([randint(0,n*10) for i in range(n)])
ens2 = sorted([randint(0,n*10) for i in range(n)])
deb = time()
inter1 = intersection(ens1, ens2)
print("Temps pour intersection1 :", time()-deb)
deb = time()
inter2 = intersection2(ens1, ens2)
print("Temps pour intersection2 :", time()-deb)

    

File: mp2i/Evaluations/DS/DS3/ensemble.ml
let rec est_ensemble l =
  match l with
  | [] -> true
  | h::[] -> true
  | h1::h2::t -> h1<h2 && est_ensemble (h2::t);;


let rec appartient x l =
  match l with
  | [] -> false
  | h::t -> if h=x then true else
    if h>x then false else appartient x t;;


let rec union l1 l2 =
  match l1, l2 with
  | [], l2 -> l2
  | l1, [] -> l1
  | h1::t1, h2::t2 -> if h1<h2 then h1::(union t1 l2) else 
    if h1=h2 then h1::(union t1 t2) else 
      h2::(union l1 t2);;

let rec intersection ens1 ens2 =
  match ens1, ens2 with
  | [], ens2 -> []
  | ens1, [] -> []
  | e1::r1, e2::r2 -> if e1=e2 then e1::(intersection r1 r2) else
    if e1<e2 then intersection r1 ens2 else
      intersection r2 ens1;;